packages/squiggle-lang/src/public/SqProject/SqModule.ts

```diff
@@ -0,0 +1,180 @@
+import { parse } from "../../ast/parse.js";
+import { AST, LocationRange } from "../../ast/types.js";
+import { Env } from "../../dists/env.js";
+import { errMap, result } from "../../utility/result.js";
+import {
+  SqCompileError,
+  SqError,
+  SqImportError,
+  SqOtherError,
+} from "../SqError.js";
+import { SqLinker } from "../SqLinker.js";
+import { ModulePointer, ProjectState } from "./ProjectState.js";
+import { SqModuleOutput } from "./SqModuleOutput.js";
+import { getHash } from "./utils.js";
+
+export type Import = {
+  name: string;
+  hash: string | undefined;
+  variable: string;
+  location: LocationRange;
+};
+
+export function importToPointer(importBinding: Import): ModulePointer {
+  return {
+    name: importBinding.name,
+    hash: importBinding.hash,
+  };
+}
+
+type ImportModules =
+  | {
+      type: "loaded";
+      value: Record<string, SqModule>;
+    }
+  | {
+      type: "loading";
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+type ImportOutputs =
+  | {
+      type: "loaded";
+      value: Record<string, SqModuleOutput>;
+    }
+  | {
+      type: "loading"; // or running - i.e. ImportModules might be loaded, but outputs don't exist yet
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+export class SqModule {
+  name: string;
+  code: string;
+  // key is module name, value is hash
+  pins: Record<string, string>;
+
+  private _ast?: result<AST, SqError>;
+
+  constructor(params: {
+    name: string;
+    code: string;
+    pins?: Record<string, string>;
+  }) {
+    this.name = params.name;
+    this.code = params.code;
+    this.pins = params.pins ?? {};
+  }
+
+  // For now, parsing is done lazily but synchronously and on happens on the
+  // main thread. Parsing is usually fast enough and this makes the
+  // implementation simpler.
+  ast() {
+    if (!this._ast) {
+      this._ast = errMap(
+        parse(this.code, this.name),
+        (e) => new SqCompileError(e)
+      );
+    }
+    return this._ast;
+  }
+
+  getImports(linker: SqLinker): Import[] {
+    const ast = this.ast();
+    if (!ast.ok) {
+      return [];
+    }
+    const program = ast.value;
+
+    const resolvedImports: Import[] = [];
+
+    for (const [file, variable] of program.imports) {
+      const name = linker.resolve(file.value, this.name);
+      resolvedImports.push({
+        name,
+        hash: this.pins[name],
+        variable: variable.value,
+        // TODO - this is used for errors, but we should use the entire import statement;
+        // To fix this, we need to represent each import statement as an AST node.
+        location: file.location,
+      });
+    }
+
+    return resolvedImports;
+  }
+
+  hash(): string {
+    return getHash(
+      `module/${this.name}/` +
+        JSON.stringify({
+          name: this.name,
+          code: this.code,
+          pins: this.pins,
+        })
+    );
+  }
+
+  // Helper methods
```

Comment:
The methods below are somewhat awkward and their return types are more ad-hoc than I'd like (is it enough to return `{ type: "loading" }`, or do we need the details about which imports are loading? for now it's enough, but hard to tell if it's going to stay this way).

By this point I was just looking for _some_ solution for "I have a module and need to check if it's ready for running" and similar problems, and I was moving the code around until it ended up here.

Passing the `state` as a parameter is also weird; these are multi-methods on "module + state". Maybe they belong in `ProjectState` and not on this class.

But there are so many other small improvements here to be made that this is pretty low on the list of priorities.

Reply:
(moved to the comment in source)